### ISSUE TITLE
SDPA DSL engines: torch glue ops run on torch's current stream, not the handle stream

### ISSUE THREAD

**Issue by YangXu1990uiuc** [NVIDIA/cudnn-frontend#510]
Two coupled problems in the SDPA DSL engine lowering layer
(`python/cudnn/sdpa/fwd/api_dsl.py`, `python/cudnn/sdpa/bwd/api_dsl.py`).

**Not in scope — already clean.** The discovery/check path was made
framework-neutral in #502: `family_for()`, the facts vocabulary, and
`Capabilities.mismatch()` reach no torch and no DSL, and
`test/python/test_import_boundaries.py` pins that. The two remaining torch
helpers in `graph_analyzer` (`_torch_cuda_device`, `to_torch_dtype`) are
deliberate, function-local, and documented as *the lowering boundary*. This
issue is about what happens **after** that boundary.

---

## 1. Stream ordering

`engines.py` launches the kernel on the **execute-time handle's stream**:

```python
# python/cudnn/sdpa/fwd/engines.py
# Stream from the execute-time handle (raw CUstream int, the
# ExecutionContext's stream); None keeps the default stream.
current_stream=_cuda_driver().CUstream(stream) if stream is not None else None,
```

`api_dsl` honours that in exactly one place — the pre-kernel amax reset — and
the comment there states the rule correctly:

```python
# Same-stream ordering as MXFP8: the resets must precede the kernel's
# atomicMax on the launch stream, not on torch's current stream.
with _torch_stream_context(current_stream, device):
    amax_s_buf.zero_()
    amax_o_buf.zero_()
```

Every other torch-side op in these paths is outside such a context, so it runs
on **torch's current stream** and races with the kernel.

**Post-kernel** — torch op may run before the kernel produces the data:

| method | op |
|---|---|
| `SdpaFwdDslSm100.execute` | `O_view.copy_(O_scratch)` |
| `SdpaFwdDslSm100._execute_mxfp8` | `O_view.copy_(O)` |
| `SdpaFwdDslSm100._execute_fp8` | `O_view.copy_(O)`, `amax_o_buf.div_(...)` |
| `SdpaFwdDslSm120.execute` | `o_view.copy_(o_scratch)` |
| `SdpaFwdDslSm120._execute_fp8` | `o_view.copy_(o_scratch)`, `amax_o_buf.div_(...)` |

**Pre-kernel** — kernel may read a buffer before the torch op fills it.
`SdpaFwdDslSm100._execute_thd` and `SdpaFwdDslSm120._execute_thd`: the
`cu_seqlens`/`meta` staging (`slq.copy_`, `meta[0:b].copy_`, `cu_q[0:1].zero_`)
and the `o_buf` / `LSE` / `sinks_t` zeroing.

**Impact.** Fires only when the handle's stream differs from torch's current
stream — the normal case for a framework that sets a handle stream, and exactly
the configuration the FROST stream-handle work enabled. Symptom: intermittent
wrong `O` / `Amax_O`, load-dependent, invisible under a synchronising debugger.

**Why the tests do not catch it.** They execute on torch's current stream, so
the two streams coincide and the race cannot fire. A regression test needs a
handle whose stream is a non-default stream.

## 2. torch is the glue layer, which is why (1) keeps happening

Per-execute device work done through torch in `sdpa/fwd/api_dsl.py`:

```
14 .zero_()   11 .copy_()   4 torch.cumsum   2 .div_()   1 torch.cat
```

(Allocations are mostly not per-execute: `_dummy` caches, and #484 already took
`torch.empty` out of execute.)

Concretely, a per-tensor FP8 dense execute launches **three glue kernels around
one attention kernel** — `amax_s.zero_()`, `amax_o.zero_()`, `amax_o.div_()` —
plus an `O` `copy_` when the output needs restaging. THD adds four `cumsum`s and
the `cu_seqlens` staging. Against the measured FE execute overhead these are not
noise.

Every one of those ops carries torch's own stream binding. That is the mechanism
behind section 1: the leak is not "someone forgot a `with` block", it is that the
glue lives in a framework whose ops bind to a different stream than the kernel.
Each newly added glue op re-opens the hole.

Related, not part of this issue: the per-tensor FP8 paths read scalar descales
host-side (`float(t.reshape(-1)[0].item())`), so they cannot be CUDA-graph
captured regardless of the fix — a `.item()` is illegal during capture.

## Fix options

**(a) Wrap.** Put every op above in the existing
`_torch_stream_context(current_stream, device)`. Cheap and local. The seam
remains: correctness stays contingent on remembering the wrapper next time.

**(b) Remove the glue.** For the FP8 paths specifically:

- `amax_o_buf.div_(scale_o)` — the kernel already carries the `o_scale_fused`
  scalar; folding the division into the epilogue removes the launch *and* the
  race together.
- the two `zero_()`s — either initialise in-kernel, or `cuMemsetD32Async` on the
  launch stream (`cuda.bindings` is already a lowering dependency).

(b) fixes correctness and overhead at once and does not depend on future authors
remembering a rule. (a) is the right immediate step if (b) is not scheduled.

Suggest doing (a) across all the sites listed, and (b) for the FP8 amax ops.

<sub>note to self: claude::304e9e55-1db7-4285-967f-001cb21032f3 — "审计 Frost Python DSL 引擎调用流程"<br>cwd /home/scratch.yanxu_libs/cudnn_frontend · workspace /home/scratch.yanxu_libs/fe_sm120</sub>

**Comment by YangXu1990uiuc** [comment]
Measured the glue cost on an RTX PRO 6000 Blackwell Server Edition (sm120,
cuDNN 9.25.0.15), per-tensor FP8 SDPA forward, CUPTI device time via
`torch.profiler`, 2 s clock warm-up per configuration and both measurement
orders (agreement within 0.5%).

Per `graph.execute()`, the DSL engine launches **4 kernels**; the backend's
native fp8 fprop on the same op launches **2**:

| | kernels | non-attention device time |
|---|---|---|
| `sdpa_fwd_prefill_sm120_fp8` | 4 | **4.5 – 5.1 us** |
| `cudnn_generated_fort_native_sdpa_sm120` | 2 | 0.7 – 1.0 us |

The three extra launches are exactly the torch glue listed above —
`amax_s.zero_()`, `amax_o.zero_()`, `amax_o.div_(scale_o)`.

The ~4.3 us delta is **constant across all 12 shapes measured** (s=1k..8k,
b=1/4, MHA and GQA 4:1), i.e. it is fixed per-execute overhead, not something
that amortises. Relative to the attention kernel it is 17% at s=1024 and 0.4%
at s=8192.

That is the argument for fix option (b) over (a): folding `amax_o /= scale_o`
into the epilogue and initialising the amax buffers in-kernel (or via
`cuMemsetD32Async` on the launch stream) removes measurable time *and* closes
the stream hole, rather than only closing the hole.

**Comment by 0z5a** [comment]
Hi @YangXu1990uiuc, I'd like to take the stream-ordering correctness slice here.

I'll first cover the remaining torch glue sites with the existing _torch_stream_context(current_stream, device) and add regression tests where the execute-time handle uses a non-default stream different from torch's current stream. I'll exercise both pre-kernel staging/zeroing and post-kernel copy/scale paths, using explicit event dependencies rather than global synchronization so the race is actually observable.

I can validate this on L20s, including repeated execution and device/stream isolation. I'd keep the more invasive FP8 glue removal (cuMemsetD32Async / epilogue folding) as a separate follow-up once the correctness fix is landed.

**Comment by 0z5a** [comment]
Audit + coverage slice is up: #1149.

Short version: every site your report lists is already protected. `engines.py::_execute` / `_execute_by_tensor` wrap the whole adapter call (they are the only callers of `api.execute`), so the copy/zero/scale calls sit inside that outer stream context; the SM100 FP8/MXFP8 arms and the prepared dense path additionally carry local contexts. Per-site statuses and caller chains are in the PR.

So the PR is coverage, not a fix: the architecture-independent cases were trapped behind the Blackwell gate in `sdpa/frost/test_sdpa_stream_ordering.py` (5/5 skipped on an L20), and the boundaries were not covered anywhere. `sdpa/test_stream_context_contract.py` now runs them anywhere (10 passed on 2 L20s) — enter/restore, exception and nesting restore, `verify_current` vs the raw fast path, the `None`/0/1/2 sentinels, two-device isolation, and a structural pin of the outer guard. The engine-level cases keep their gate.

The sentinel branch is probed but unchanged, per the "no guessing fix" rule. The ordering probe's sensitivity is shown by a test-local mutation (context -> no-op): it reads the poison from 8e8 spin cycles up, the real context is correct at 1e9.
